Consistent styling for Error Page (404 page)

Vall3y opened the pull request:
I have applied the same style from `<InANutShell />` to the error page. To do so I have created a shared styles folder under web_modules/app, similarly to how it is done in this [CSS Modules demo](https://css-modules.github.io/webpack-demo/) (See StyleVariantA.css)

I didn't want to refactor any more code before getting some feedback if this style is acceptable, let me know what you think

![image](https://cloud.githubusercontent.com/assets/325644/14401266/12277a0c-fe14-11e5-9989-3b315ff85663.png)

marcustisater commented:
Thanks for the PR. I will take a look.

jeddy3 commented:
@Vall3y Thanks for getting stuck in :)

I'm easy if you guys want to go down the route of *abstracting styles out of components*, but in my experience it actually make maintenance more difficult by increasing odds of *side-effects* and the cognitive load on developers. I intentionally avoided these kind of abstractions when I first building the homepage. I added, at the time, a small note the [CSS Processors](https://github.com/postcss/postcss.org/blob/master/docs/react.md#css-processors) section of the `react` guide to help explain why:

"It is possible to [compose selectors with CSS Modules](https://github.com/css-modules/css-modules#composition). However, do _not_ [compose class names from other CSS Modules](https://github.com/css-modules/css-modules#dependencies) as dependencies make it more difficult to [managed side effects](http://philipwalton.com/articles/side-effects-in-css/) in CSS."

That's not to say you can't use composition *inside* of a component, for example [inside of the `Social` component](https://github.com/postcss/postcss.org/blob/master/web_modules/Social/index.css#L24-L50). I think there's less for a developer to mentally model if these instances and the side-effects are limited to within the component.

Anyhow, like I said, you guys can take this down which ever route is most comfortable for you. I just figured you'd appreciate knowing the original rationale behind not using these type of abstractions :)

marcustisater commented:
I agree with @jeddy3, that's a very good point indeed. I'm speaking as a novice CSS Modules user and in my opinion composing selectors outside components is very much like a bad practiced "extend" creating more unnecessary complications rather then good use. 

Irrelevant example but after being a very long Sass time user I know such things are tempting to reduce writing but sorry to say I have always disliked it because usually it creates more side effects.

I'd say refactor this, let's not abstract styles out of components.

marcustisater commented:
Irrelevant(sorry): @jeddy3 - did you get my issue message? You should jump into [gitter](https://gitter.im/postcss/postcss.org)! :+1:

marcustisater commented:
@Vall3y - Would you like to refactor this from the feedback or close for new PR?

Vall3y commented:
Thanks for the clarification, I do appreciate that. I can refactor this PR

So I understand that the preferred method is for the React class to import and apply the common styles itself, rather than for the css classes to be composed. Is this correct?

In the [post](http://philipwalton.com/articles/extending-styles#web-component-considerations) @jeddy has linked, Phillip Walton actually seems okay with css classes extending others, even though I do understand why your solution might be better though

thangngoc89 commented:
IMHO, composing is a bad idea. I usually refactor React component to avoid that

jeddy3 commented:
> So I understand that the preferred method is for the React class to import and apply the common styles itself, rather than for the css classes to be composed. Is this correct?

I think the thing to do here is not spend time abstracting out common styles, but rather spend that time designing and building the styles for an unique looking error page. Something that clearly distinguishes it from a "normal" page, so that it better communicates to the user that something has gone wrong.

> The other solution would be to have the each component have a duplicate of the style statements

We don't have to worry about this problem then... and we end up with a nice looking error page :)

Vall3y commented:
I have reverted the abstraction changes, and added some of its own style. I've also added a redirect to homepage message. Let me know if this is okay

![image](https://cloud.githubusercontent.com/assets/325644/14407054/d47a85f4-fec3-11e5-90ae-8b27ee6384ff.png)

jeddy3 commented:
LGTM :)

marcustisater commented:
LGTM 

Thanks everyone for the discussion of  abstracting styles, got me thinking as well.
